lib/theory/src/Theory/Constraint/System/JSON.hs

```diff
@@ -0,0 +1,394 @@
+  deriving (Show)
+
+-- fact in a JSON graph node
+data JSONGraphNodeFact = JSONGraphNodeFact {
+    jgnFactId    :: String
+  , jgnFactTag   :: String  -- ProtoFact, FreshFact, OutFact, InFact, KUFact, KDFact, DedFact
+  , jgnFactName  :: String  -- Fr, Out, In, !KU, ...
+  , jgnFactMult  :: String  -- "!" = persistent, "" = linear
+  , jgnFactTerms :: [JSONGraphNodeTerm]
+  , jgnFactShow  :: String
+} deriving (Show)
+
+-- meta data of JSON graph node
+data JSONGraphNodeMetadata = JSONGraphNodeMetadata {
+    jgnPrems :: [JSONGraphNodeFact] -- all could be done as optional
+  , jgnActs  :: [JSONGraphNodeFact]
+  , jgnConcs :: [JSONGraphNodeFact]
+} deriving (Show)
+
+-- node of a JSON graph
+data JSONGraphNode = JSONGraphNode {
+    jgnId :: String
+  , jgnType :: String
+  , jgnLabel :: String
+  , jgnMetadata :: Maybe JSONGraphNodeMetadata
+} deriving (Show)
+
+-- optional fields are not handled correctly with automatically derived instances
+-- hence, we have our own here
+instance FromJSON JSONGraphNode where
+    parseJSON = withObject "JSONGraphNode" $ \o -> JSONGraphNode
+        <$> o .: "jgnId"
+        <*> o .: "jgnType"
+        <*> o .: "jgnLabel"
+        <*> o .:? "jgnMetadata"
+
+instance ToJSON JSONGraphNode where
+    toJSON (JSONGraphNode jgnId jgnType jgnLabel jgnMetadata) = object $ catMaybes
+        [ ("jgnId" .=) <$> pure jgnId      
+        , ("jgnType" .=) <$> pure jgnType
+        , ("jgnLabel" .=) <$> pure jgnLabel
+        , ("jgnMetadata" .=) <$> jgnMetadata ]
+
+-- edge of a JSON graph
+data JSONGraphEdge = JSONGraphEdge {
+    jgeSource :: String
+  , jgeRelation :: String
+  , jgeTarget :: String
+--  , jgeDirected :: Maybe Bool
+--  , jgeLabel :: Maybe String
+} deriving (Show)
+
+-- JSON graph
+data JSONGraph = JSONGraph {
+    jgDirected :: Bool
+  , jgType :: String
+  , jgLabel :: String
+  , jgNodes :: [JSONGraphNode]
+  , jgEdges :: [JSONGraphEdge]
+--  , jgmetadata :: JSONGraphMetadata 
+} deriving (Show)
+
+-- collection of JSON graphs
+data JSONGraphs = JSONGraphs {
+    graphs :: [JSONGraph]
+} deriving (Show)
+
+-- derive ToJSON and FromJSON 
+concat <$> mapM (deriveJSON defaultOptions) [''JSONGraphNodeTerm, ''JSONGraphNodeFact, ''JSONGraphNodeMetadata, ''JSONGraphEdge, ''JSONGraph, ''JSONGraphs]
+
+------------------------------------------------------------------------------
+-- Generation of JSON text from JSON graphs                                 --
+------------------------------------------------------------------------------
+
+-- flatten out pretty printed facts from prettyLNFact etc.
+cleanString :: [Char] -> [Char]
+cleanString [] = []
+cleanString (' ':'\n':' ':xs) = cleanString (' ':xs)
+cleanString ('\n':xs) = cleanString xs
+cleanString (' ':' ':xs) = cleanString (' ':xs)
+cleanString (c:xs) = (c:cleanString xs)
+
+-- convert output of pretty print functions to string
+pps :: Doc -> String
+pps d = cleanString $ render d
+
+-- encodePretty encodes '<' as "\u003c" and '>' as "\u003e"
```

Comment:
Hmm... could it be that your JSON parser has trouble parsing Unicode escapes? The \u3e is a valid escape.